File: gitgov/src/components/control_plane/dashboard-helpers/first-governed-repo-setup.ts
export type FirstGovernedRepoSetupGoal =
  | 'govern_release'
  | 'generate_audit_evidence'
  | 'standardize_workflows'
  | 'assess_governance_gaps'

export type FirstGovernedRepoSetupStatus = 'draft' | 'ready' | 'blocked' | 'completed'
export type FirstGovernedRepoPolicyPreset = 'audit-only' | 'moderate' | 'strict'
export type FirstGovernedRepoProvider = 'github' | 'jira' | 'jenkins' | 'sonarqube' | 'render' | 'vercel'
export type FirstGovernedRepoModule =
  | 'traceability'
  | 'github-evidence'
  | 'release-readiness'
  | 'quality-gates'
  | 'evidence-packets'
  | 'formal-approval'
export type FirstGovernedRepoGateReadiness = 'needs_repo' | 'needs_preview' | 'baseline_ready'

export interface FirstGovernedRepoOption<T extends string> {
  id: T
  label: string
  description: string
  required?: boolean
}

export interface FirstGovernedRepoSetupBaseline extends Record<string, unknown> {
  version: 1
  policy_workflow_preview_acknowledged: boolean
  gate_readiness: FirstGovernedRepoGateReadiness
  setup_summary: {
    repository_full_name: string
    default_branch: string
    goal: FirstGovernedRepoSetupGoal
    policy_preset: FirstGovernedRepoPolicyPreset
    provider_count: number
    module_count: number
    github_selected: boolean
    policy_workflow_preview_acknowledged: boolean
  }
  action_center_gaps: string[]
  first_result: {
    status: 'ready_for_advisory_gate' | 'needs_setup'
    deployment_gate_mode: 'advisory'
    cta: 'simulate_deployment_gate'
    evidence_contract: {
      repo: string
      branch: string
      providers: FirstGovernedRepoProvider[]
      modules: FirstGovernedRepoModule[]
    }
  }
}

export interface FirstGovernedRepoSetupDraft {
  status: FirstGovernedRepoSetupStatus
  goal: FirstGovernedRepoSetupGoal
  repository_full_name: string
  default_branch: string
  selected_providers: FirstGovernedRepoProvider[]
  selected_modules: FirstGovernedRepoModule[]
  policy_preset: FirstGovernedRepoPolicyPreset
  baseline: FirstGovernedRepoSetupBaseline
}

export interface FirstGovernedRepoSetupValidation {
  ready: boolean
  gateReadiness: FirstGovernedRepoGateReadiness
  gaps: string[]
  errors: string[]
}

export const FIRST_GOVERNED_REPO_GOAL_OPTIONS: Array<FirstGovernedRepoOption<FirstGovernedRepoSetupGoal>> = [
  {
    id: 'govern_release',
    label: 'Govern release',
    description: 'Prepare the first repo for advisory deployment gate evidence.',
  },
  {
    id: 'generate_audit_evidence',
    label: 'Audit evidence',
    description: 'Prioritize traceable evidence packets and provider proof.',
  },
  {
    id: 'standardize_workflows',
    label: 'Workflow baseline',
    description: 'Standardize reviewed workflow and policy installation steps.',
  },
  {
    id: 'assess_governance_gaps',
    label: 'Gap assessment',
    description: 'Expose missing controls before gate simulation.',
  },
]

export const FIRST_GOVERNED_REPO_PROVIDER_OPTIONS: Array<FirstGovernedRepoOption<FirstGovernedRepoProvider>> = [
  {
    id: 'github',
    label: 'GitHub',
    description: 'Repo, PR, status, and webhook evidence.',
    required: true,
  },
  {
    id: 'jira',
    label: 'Jira',
    description: 'Ticket traceability and release context.',
  },
  {
    id: 'jenkins',
    label: 'Jenkins',
    description: 'Pipeline run and build outcome evidence.',
  },
  {
    id: 'sonarqube',
    label: 'SonarQube',
    description: 'Quality gate evidence inferred from pipeline signals.',
  },
  {
    id: 'render',
    label: 'Render',
    description: 'Deployment target evidence when configured later.',
  },
  {
    id: 'vercel',
    label: 'Vercel',
    description: 'Deployment target evidence when configured later.',
  },
]

export const FIRST_GOVERNED_REPO_MODULE_OPTIONS: Array<FirstGovernedRepoOption<FirstGovernedRepoModule>> = [
  {
    id: 'traceability',
    label: 'Traceability',
    description: 'Ticket-to-code coverage.',
    required: true,
  },
  {
    id: 'release-readiness',
    label: 'Release readiness',
    description: 'Advisory readiness score input.',
    required: true,
  },
  {
    id: 'evidence-packets',
    label: 'Evidence packets',
    description: 'Downloadable audit packet proof.',
    required: true,
  },
  {
    id: 'quality-gates',
    label: 'Quality gates',
    description: 'CI quality pass/fail evidence.',
  },
  {
    id: 'github-evidence',
    label: 'GitHub evidence',
    description: 'PR review and check-run signal depth.',
  },
  {
    id: 'formal-approval',
    label: 'Formal approval',
    description: 'Release approval record before hard enforcement.',
  },
]

export const FIRST_GOVERNED_REPO_POLICY_PRESET_OPTIONS: Array<FirstGovernedRepoOption<FirstGovernedRepoPolicyPreset>> = [
  {
    id: 'audit-only',
    label: 'Audit only',
    description: 'Collect evidence without advisory warnings.',
  },
  {
    id: 'moderate',
    label: 'Moderate',
    description: 'Recommended advisory deployment gate baseline.',
  },
  {
    id: 'strict',
    label: 'Strict',
    description: 'Prepare for future blocking controls.',
  },
]

export const DEFAULT_FIRST_GOVERNED_REPO_SETUP: FirstGovernedRepoSetupDraft = {
  status: 'draft',
  goal: 'govern_release',
  repository_full_name: '',
  default_branch: 'main',
  selected_providers: ['github'],
  selected_modules: ['traceability', 'release-readiness', 'evidence-packets'],
  policy_preset: 'moderate',
  baseline: buildFirstGovernedRepoSetupBaseline({
    repository_full_name: '',
    default_branch: 'main',
    goal: 'govern_release',
    selected_providers: ['github'],
    selected_modules: ['traceability', 'release-readiness', 'evidence-packets'],
    policy_preset: 'moderate',
    policyWorkflowPreviewAcknowledged: false,
  }),
}

function dedupeAllowed<T extends string>(values: unknown, allowed: readonly T[], fallback: T[]): T[] {
  if (!Array.isArray(values)) return fallback
  const allowedSet = new Set(allowed)
  const seen = new Set<T>()
  const result: T[] = []
  for (const value of values) {
    if (typeof value !== 'string') continue
    const trimmed = value.trim() as T
    if (!allowedSet.has(trimmed) || seen.has(trimmed)) continue
    seen.add(trimmed)
    result.push(trimmed)
  }
  return result.length > 0 ? result : fallback
}

export function isFirstGovernedRepoNameValid(repositoryFullName: string): boolean {
  const value = repositoryFullName.trim()
  if (value.length === 0 || value.length > 200 || /\s/.test(value)) return false
  const parts = value.split('/')
  if (parts.length !== 2) return false
  return parts.every((part) => /^[A-Za-z0-9_.-]{1,100}$/.test(part) && !part.startsWith('.') && !part.endsWith('.'))
}

export function buildFirstGovernedRepoSetupBaseline(input: {
  repository_full_name: string
  default_branch: string
  goal: FirstGovernedRepoSetupGoal
  selected_providers: FirstGovernedRepoProvider[]
  selected_modules: FirstGovernedRepoModule[]
  policy_preset: FirstGovernedRepoPolicyPreset
  policyWorkflowPreviewAcknowledged: boolean
}): FirstGovernedRepoSetupBaseline {
  const repositoryFullName = input.repository_full_name.trim()
  const defaultBranch = input.default_branch.trim() || 'main'
  const providers = input.selected_providers.includes('github')
    ? input.selected_providers
    : (['github', ...input.selected_providers] as FirstGovernedRepoProvider[])
  const modules = input.selected_modules
  const repoReady = isFirstGovernedRepoNameValid(repositoryFullName)
  const previewReady = input.policyWorkflowPreviewAcknowledged
  const gateReadiness: FirstGovernedRepoGateReadiness =
    repoReady && previewReady ? 'baseline_ready' : repoReady ? 'needs_preview' : 'needs_repo'
  const gaps: string[] = []
  if (!repoReady) gaps.push('repository_full_name')
  if (!previewReady) gaps.push('policy_workflow_preview')
  if (!modules.includes('quality-gates')) gaps.push('quality_gate_evidence')
  if (!modules.includes('formal-approval')) gaps.push('formal_approval_policy')

  return {
    version: 1,
    policy_workflow_preview_acknowledged: previewReady,
    gate_readiness: gateReadiness,
    setup_summary: {
      repository_full_name: repositoryFullName,
      default_branch: defaultBranch,
      goal: input.goal,
      policy_preset: input.policy_preset,
      provider_count: providers.length,
      module_count: modules.length,
      github_selected: providers.includes('github'),
      policy_workflow_preview_acknowledged: previewReady,
    },
    action_center_gaps: gaps,
    first_result: {
      status: gateReadiness === 'baseline_ready' ? 'ready_for_advisory_gate' : 'needs_setup',
      deployment_gate_mode: 'advisory',
      cta: 'simulate_deployment_gate',
      evidence_contract: {
        repo: repositoryFullName,
        branch: defaultBranch,
        providers,
        modules,
      },
    },
  }
}

export function normalizeFirstGovernedRepoSetupDraft(input?: Partial<FirstGovernedRepoSetupDraft> | null): FirstGovernedRepoSetupDraft {
  const allowedGoals = FIRST_GOVERNED_REPO_GOAL_OPTIONS.map((option) => option.id)
  const allowedProviders = FIRST_GOVERNED_REPO_PROVIDER_OPTIONS.map((option) => option.id)
  const allowedModules = FIRST_GOVERNED_REPO_MODULE_OPTIONS.map((option) => option.id)
  const allowedPresets = FIRST_GOVERNED_REPO_POLICY_PRESET_OPTIONS.map((option) => option.id)
  const selectedProviders = dedupeAllowed(
    input?.selected_providers,
    allowedProviders,
    DEFAULT_FIRST_GOVERNED_REPO_SETUP.selected_providers,
  )
  const providers = selectedProviders.includes('github')
    ? selectedProviders
    : (['github', ...selectedProviders] as FirstGovernedRepoProvider[])
  const modules = dedupeAllowed(
    input?.selected_modules,
    allowedModules,
    DEFAULT_FIRST_GOVERNED_REPO_SETUP.selected_modules,
  )
  const goal = allowedGoals.includes(input?.goal as FirstGovernedRepoSetupGoal)
    ? (input?.goal as FirstGovernedRepoSetupGoal)
    : DEFAULT_FIRST_GOVERNED_REPO_SETUP.goal
  const policyPreset = allowedPresets.includes(input?.policy_preset as FirstGovernedRepoPolicyPreset)
    ? (input?.policy_preset as FirstGovernedRepoPolicyPreset)
    : DEFAULT_FIRST_GOVERNED_REPO_SETUP.policy_preset
  const previewAck =
    Boolean(input?.baseline?.policy_workflow_preview_acknowledged) ||
    input?.baseline?.gate_readiness === 'baseline_ready'
  const baseline = buildFirstGovernedRepoSetupBaseline({
    repository_full_name: input?.repository_full_name ?? '',
    default_branch: input?.default_branch ?? 'main',
    goal,
    selected_providers: providers,
    selected_modules: modules,
    policy_preset: policyPreset,
    policyWorkflowPreviewAcknowledged: previewAck,
  })

  return {
    status: baseline.gate_readiness === 'baseline_ready' ? 'ready' : 'draft',
    goal,
    repository_full_name: input?.repository_full_name?.trim() ?? '',
    default_branch: input?.default_branch?.trim() || 'main',
    selected_providers: providers,
    selected_modules: modules,
    policy_preset: policyPreset,
    baseline,
  }
}

export function validateFirstGovernedRepoSetupDraft(
  draft: FirstGovernedRepoSetupDraft,
): FirstGovernedRepoSetupValidation {
  const baseline = buildFirstGovernedRepoSetupBaseline({
    repository_full_name: draft.repository_full_name,
    default_branch: draft.default_branch,
    goal: draft.goal,
    selected_providers: draft.selected_providers,
    selected_modules: draft.selected_modules,
    policy_preset: draft.policy_preset,
    policyWorkflowPreviewAcknowledged: draft.baseline.policy_workflow_preview_acknowledged,
  })
  const errors: string[] = []
  if (!isFirstGovernedRepoNameValid(draft.repository_full_name)) {
    errors.push('Repository must use owner/repo format.')
  }
  if (!draft.selected_providers.includes('github')) {
    errors.push('GitHub is required for the first governed repo.')
  }
  if (draft.default_branch.trim().length === 0 || /\s/.test(draft.default_branch)) {
    errors.push('Default branch must not be empty or contain whitespace.')
  }

  return {
    ready: baseline.gate_readiness === 'baseline_ready' && errors.length === 0,
    gateReadiness: baseline.gate_readiness,
    gaps: baseline.action_center_gaps,
    errors,
  }
}

File: gitgov/gitgov-server/src/handlers/first_governed_repo_setup.rs
// ============================================================================
// FIRST GOVERNED REPO SETUP
// ============================================================================

const FIRST_GOVERNED_REPO_BASELINE_MAX_BYTES: usize = 24 * 1024;
const FIRST_GOVERNED_REPO_STATUSES: &[&str] = &["draft", "ready", "blocked", "completed"];
const FIRST_GOVERNED_REPO_GOALS: &[&str] = &[
    "govern_release",
    "generate_audit_evidence",
    "standardize_workflows",
    "assess_governance_gaps",
];
const FIRST_GOVERNED_REPO_POLICY_PRESETS: &[&str] = &["audit-only", "moderate", "strict"];
const FIRST_GOVERNED_REPO_PROVIDER_IDS: &[&str] =
    &["github", "jira", "jenkins", "sonarqube", "render", "vercel"];
const FIRST_GOVERNED_REPO_MODULE_IDS: &[&str] = &[
    "traceability",
    "github-evidence",
    "release-readiness",
    "quality-gates",
    "evidence-packets",
    "formal-approval",
];
const FIRST_GOVERNED_REPO_SECRET_MARKERS: &[&str] = &[
    "authorization:",
    "bearer ",
    "api_token",
    "api_token=",
    "gitgov_api_key",
    "gitgov_api_key=",
    "jira_api_token",
    "jira_api_token=",
    "sonar_token",
    "sonar_token=",
    "jenkins_api_token",
    "jenkins_api_token=",
    "render_api_key",
    "render_api_key=",
    "vercel_token",
    "vercel_token=",
    "vck_",
    "gho_",
    "github_pat_",
    "atatt",
];

fn first_governed_repo_scope_error_message(error: OrgScopeError) -> &'static str {
    match error {
        OrgScopeError::BadRequest => "org_name is required for global admin keys",
        OrgScopeError::NotFound => "org_name was not found",
        OrgScopeError::Forbidden => "API key is not authorized for that org_name",
        OrgScopeError::Internal => "Internal database error",
    }
}

fn first_setup_trimmed_or_default(value: &str, default: &str) -> String {
    let trimmed = value.trim();
    if trimmed.is_empty() {
        default.to_string()
    } else {
        trimmed.to_string()
    }
}

fn first_setup_valid_repo_part(part: &str) -> bool {
    !part.is_empty()
        && part.len() <= 100
        && !part.starts_with('.')
        && !part.ends_with('.')
        && part
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'-' | b'_' | b'.'))
}

fn first_setup_repo_format_valid(repository_full_name: &str) -> bool {
    let trimmed = repository_full_name.trim();
    if trimmed.len() > 200 || trimmed.contains(char::is_whitespace) {
        return false;
    }
    let parts: Vec<&str> = trimmed.split('/').collect();
    parts.len() == 2
        && first_setup_valid_repo_part(parts[0])
        && first_setup_valid_repo_part(parts[1])
}

fn first_setup_normalize_choice(
    raw: &str,
    default: &str,
    allowed: &[&str],
    field: &str,
    errors: &mut Vec<String>,
) -> String {
    let value = first_setup_trimmed_or_default(raw, default);
    if !allowed.contains(&value.as_str()) {
        errors.push(format!("{field} must be one of {}.", allowed.join(", ")));
    }
    value
}

fn first_setup_normalize_status(
    raw: Option<&str>,
    default: &str,
    errors: &mut Vec<String>,
) -> String {
    let value = raw
        .map(|item| first_setup_trimmed_or_default(item, default))
        .unwrap_or_else(|| default.to_string());
    if !FIRST_GOVERNED_REPO_STATUSES.contains(&value.as_str()) {
        errors.push(format!(
            "status must be one of {}.",
            FIRST_GOVERNED_REPO_STATUSES.join(", ")
        ));
    }
    value
}

fn first_setup_normalize_string_array(
    values: &[String],
    field: &str,
    allowed: &[&str],
    default_values: &[&str],
    require_github: bool,
    errors: &mut Vec<String>,
) -> Vec<String> {
    let source: Vec<String> = if values.is_empty() {
        default_values.iter().map(|value| value.to_string()).collect()
    } else {
        values.to_vec()
    };
    if source.len() > allowed.len() {
        errors.push(format!("{field} has too many values."));
    }

    let allowed_set: HashSet<&str> = allowed.iter().copied().collect();
    let mut seen = HashSet::new();
    let mut normalized = Vec::new();
    for value in source {
        let trimmed = value.trim();
        if !allowed_set.contains(trimmed) {
            errors.push(format!("{field} contains unsupported value '{trimmed}'."));
            continue;
        }
        if seen.insert(trimmed.to_string()) {
            normalized.push(trimmed.to_string());
        }
    }

    if normalized.is_empty() {
        errors.push(format!("Select at least one {field} value."));
    }
    if require_github && !normalized.iter().any(|value| value == "github") {
        errors.push("selected_providers must include github.".to_string());
    }

    normalized
}

fn first_setup_contains_secret(value: &serde_json::Value) -> bool {
    match value {
        serde_json::Value::String(text) => {
            let lowered = text.to_ascii_lowercase();
            FIRST_GOVERNED_REPO_SECRET_MARKERS
                .iter()
                .any(|marker| lowered.contains(marker))
        }
        serde_json::Value::Array(items) => items.iter().any(first_setup_contains_secret),
        serde_json::Value::Object(map) => map.iter().any(|(key, value)| {
            let lowered_key = key.to_ascii_lowercase();
            FIRST_GOVERNED_REPO_SECRET_MARKERS
                .iter()
                .any(|marker| lowered_key.contains(marker))
                || first_setup_contains_secret(value)
        }),
        _ => false,
    }
}

fn first_setup_bool_field(
    object: &serde_json::Map<String, serde_json::Value>,
    field: &str,
) -> bool {
    object
        .get(field)
        .and_then(|value| value.as_bool())
        .unwrap_or(false)
}

fn first_setup_prepare_payload(
    mut payload: UpsertFirstGovernedRepoSetupRequest,
) -> Result<UpsertFirstGovernedRepoSetupRequest, Vec<String>> {
    let mut errors = Vec::new();

    payload.repository_full_name = payload.repository_full_name.trim().to_string();
    if !first_setup_repo_format_valid(&payload.repository_full_name) {
        errors.push("repository_full_name must use owner/repo format.".to_string());
    }

    payload.default_branch = first_setup_trimmed_or_default(&payload.default_branch, "main");
    if payload.default_branch.len() > 120 || payload.default_branch.contains(char::is_whitespace) {
        errors.push("default_branch must be a branch name without whitespace.".to_string());
    }

    payload.goal = first_setup_normalize_choice(
        &payload.goal,
        "govern_release",
        FIRST_GOVERNED_REPO_GOALS,
        "goal",
        &mut errors,
    );
    payload.policy_preset = first_setup_normalize_choice(
        &payload.policy_preset,
        "moderate",
        FIRST_GOVERNED_REPO_POLICY_PRESETS,
        "policy_preset",
        &mut errors,
    );
    payload.selected_providers = first_setup_normalize_string_array(
        &payload.selected_providers,
        "selected_providers",
        FIRST_GOVERNED_REPO_PROVIDER_IDS,
        &["github"],
        true,
        &mut errors,
    );
    payload.selected_modules = first_setup_normalize_string_array(
        &payload.selected_modules,
        "selected_modules",
        FIRST_GOVERNED_REPO_MODULE_IDS,
        &["traceability", "release-readiness", "evidence-packets"],
        false,
        &mut errors,
    );

    let size = serde_json::to_vec(&payload.baseline)
        .map(|bytes| bytes.len())
        .unwrap_or(usize::MAX);
    if size > FIRST_GOVERNED_REPO_BASELINE_MAX_BYTES {
        errors.push("baseline is too large.".to_string());
    }
    if first_setup_contains_secret(&payload.baseline) {
        errors.push("baseline must not contain secret-looking values.".to_string());
    }

    let mut baseline = match payload.baseline {
        serde_json::Value::Null => serde_json::Map::new(),
        serde_json::Value::Object(map) => map,
        _ => {
            errors.push("baseline must be an object.".to_string());
            serde_json::Map::new()
        }
    };

    let preview_acknowledged =
        first_setup_bool_field(&baseline, "policy_workflow_preview_acknowledged");
    let repo_ready = first_setup_repo_format_valid(&payload.repository_full_name);
    let github_ready = payload.selected_providers.iter().any(|value| value == "github");
    let gate_readiness = if repo_ready && github_ready && preview_acknowledged {
        "baseline_ready"
    } else if repo_ready && github_ready {
        "needs_preview"
    } else {
        "needs_repo"
    };

    let mut action_center_gaps = Vec::new();
    if !repo_ready {
        action_center_gaps.push("repository_full_name");
    }
    if !preview_acknowledged {
        action_center_gaps.push("policy_workflow_preview");
    }
    if !payload
        .selected_modules
        .iter()
        .any(|value| value == "quality-gates")
    {
        action_center_gaps.push("quality_gate_evidence");
    }
    if !payload
        .selected_modules
        .iter()
        .any(|value| value == "formal-approval")
    {
        action_center_gaps.push("formal_approval_policy");
    }

    let default_status = if gate_readiness == "baseline_ready" {
        "ready"
    } else {
        "draft"
    };
    let normalized_status =
        first_setup_normalize_status(payload.status.as_deref(), default_status, &mut errors);
    if normalized_status == "completed" && gate_readiness != "baseline_ready" {
        errors.push("completed status requires baseline_ready gate_readiness.".to_string());
    }
    payload.status = Some(normalized_status);

    baseline.insert("version".to_string(), json!(1));
    baseline.insert("gate_readiness".to_string(), json!(gate_readiness));
    baseline.insert(
        "setup_summary".to_string(),
        json!({
            "repository_full_name": payload.repository_full_name.clone(),
            "default_branch": payload.default_branch.clone(),
            "goal": payload.goal.clone(),
            "policy_preset": payload.policy_preset.clone(),
            "provider_count": payload.selected_providers.len(),
            "module_count": payload.selected_modules.len(),
            "github_selected": github_ready,
            "policy_workflow_preview_acknowledged": preview_acknowledged
        }),
    );
    baseline.insert(
        "action_center_gaps".to_string(),
        serde_json::Value::Array(
            action_center_gaps
                .iter()
                .map(|value| json!(value))
                .collect::<Vec<_>>(),
        ),
    );
    baseline.insert(
        "first_result".to_string(),
        json!({
            "status": if gate_readiness == "baseline_ready" { "ready_for_advisory_gate" } else { "needs_setup" },
            "deployment_gate_mode": "advisory",
            "cta": "simulate_deployment_gate",
            "evidence_contract": {
                "repo": payload.repository_full_name.clone(),
                "branch": payload.default_branch.clone(),
                "providers": payload.selected_providers.clone(),
                "modules": payload.selected_modules.clone()
            }
        }),
    );
    payload.baseline = serde_json::Value::Object(baseline);

    if errors.is_empty() {
        Ok(payload)
    } else {
        Err(errors)
    }
}


pub async fn get_first_governed_repo_setup(
    Extension(auth_user): Extension<AuthUser>,
    State(state): State<Arc<AppState>>,
    Query(query): Query<FirstGovernedRepoSetupQuery>,
) -> impl IntoResponse {
    if let Err(resp) = require_admin(&auth_user) {
        return resp.into_response();
    }

    let org_id = match resolve_and_check_org_scope(
        &state,
        auth_user.org_id.as_deref(),
        query.org_name.as_deref(),
        true,
    )
    .await
    {
        Ok(Some(org_id)) => org_id,
        Ok(None) => {
            return (
                StatusCode::BAD_REQUEST,
                Json(json!({ "error": "org_name is required for global admin keys" })),
            )
                .into_response();
        }
        Err(err) => {
            return (
                org_scope_status(err),
                Json(json!({ "error": first_governed_repo_scope_error_message(err) })),
            )
                .into_response();
        }
    };

    match state.db.get_first_governed_repo_setup(&org_id).await {
        Ok(Some(setup)) => (
            StatusCode::OK,
            Json(FirstGovernedRepoSetupResponse {
                found: true,
                setup: Some(setup),
            }),
        )
            .into_response(),
        Ok(None) => (
            StatusCode::OK,
            Json(FirstGovernedRepoSetupResponse {
                found: false,
                setup: None,
            }),
        )
            .into_response(),
        Err(e) => {
            tracing::error!(error = %e, org_id = %org_id, "Failed to load first governed repo setup");
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "error": "Internal database error" })),
            )
                .into_response()
        }
    }
}

pub async fn upsert_first_governed_repo_setup(
    Extension(auth_user): Extension<AuthUser>,
    State(state): State<Arc<AppState>>,
    Json(payload): Json<UpsertFirstGovernedRepoSetupRequest>,
) -> impl IntoResponse {
    if let Err(resp) = require_admin(&auth_user) {
        return resp.into_response();
    }

    let payload = match first_setup_prepare_payload(payload) {
        Ok(payload) => payload,
        Err(errors) => {
            return (
                StatusCode::BAD_REQUEST,
                Json(json!({ "error": "Invalid first governed repo setup", "details": errors })),
            )
                .into_response();
        }
    };

    let org_id = match resolve_and_check_org_scope(
        &state,
        auth_user.org_id.as_deref(),
        payload.org_name.as_deref(),
        true,
    )
    .await
    {
        Ok(Some(org_id)) => org_id,
        Ok(None) => {
            return (
                StatusCode::BAD_REQUEST,
                Json(json!({ "error": "org_name is required for global admin keys" })),
            )
                .into_response();
        }
        Err(err) => {
            return (
                org_scope_status(err),
                Json(json!({ "error": first_governed_repo_scope_error_message(err) })),
            )
                .into_response();
        }
    };

    match state
        .db
        .upsert_first_governed_repo_setup(&org_id, &payload, &auth_user.client_id)
        .await
    {
        Ok(record) => {
            let audit = AdminAuditLogEntry {
                id: Uuid::new_v4().to_string(),
                actor_client_id: auth_user.client_id.clone(),
                action: "upsert_first_governed_repo_setup".to_string(),
                target_type: Some("enterprise_first_governed_repo_setup".to_string()),
                target_id: Some(org_id.clone()),
                metadata: json!({
                    "org_id": org_id,
                    "run_id": record.run_id.clone(),
                    "status": record.status.clone(),
                    "goal": record.goal.clone(),
                    "repository_full_name": record.repository_full_name.clone(),
                    "default_branch": record.default_branch.clone(),
                    "policy_preset": record.policy_preset.clone(),
                    "provider_count": record.selected_providers.len(),
                    "module_count": record.selected_modules.len(),
                    "gate_readiness": record.baseline.get("gate_readiness").and_then(|value| value.as_str())
                }),
                created_at: chrono::Utc::now().timestamp_millis(),
            };
            if let Err(e) = state.db.insert_admin_audit_log(&audit).await {
                tracing::warn!(error = %e, "Failed to write admin audit log (first governed repo setup)");
            }

            (StatusCode::OK, Json(record)).into_response()
        }
        Err(e) => {
            tracing::error!(error = %e, org_id = %org_id, "Failed to save first governed repo setup");
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "error": "Internal database error" })),
            )
                .into_response()
        }
    }
}


#[cfg(test)]
mod first_governed_repo_setup_tests {
    use super::*;

    #[test]
    fn first_setup_prepare_payload_normalizes_defaults_and_gaps() {
        let payload = UpsertFirstGovernedRepoSetupRequest {
            repository_full_name: " example/repo ".to_string(),
            baseline: json!({ "policy_workflow_preview_acknowledged": true }),
            ..Default::default()
        };

        let prepared = first_setup_prepare_payload(payload).expect("valid payload");

        assert_eq!(prepared.goal, "govern_release");
        assert_eq!(prepared.default_branch, "main");
        assert_eq!(prepared.selected_providers, vec!["github"]);
        assert_eq!(
            prepared.selected_modules,
            vec!["traceability", "release-readiness", "evidence-packets"]
        );
        assert_eq!(
            prepared.baseline.get("gate_readiness").and_then(|value| value.as_str()),
            Some("baseline_ready")
        );
        assert_eq!(prepared.status.as_deref(), Some("ready"));
    }

    #[test]
    fn first_setup_prepare_payload_rejects_secret_like_baseline() {
        let payload = UpsertFirstGovernedRepoSetupRequest {
            repository_full_name: "example/repo".to_string(),
            baseline: json!({
                "notes": {
                    "jira_api_token": "redacted",
                    "token": "Bearer abc123"
                }
            }),
            ..Default::default()
        };

        let errors = first_setup_prepare_payload(payload).expect_err("secret-looking baseline");

        assert!(errors
            .iter()
            .any(|error| error.contains("secret-looking")));
    }
}
